scipy/sparse/coo.py

```diff
@@ -262,7 +262,7 @@ def _check(self):
             warn("col index array has non-integer dtype (%s) "
                     % self.col.dtype.name)
 
-        idx_dtype = get_index_dtype(maxval=max(self.shape))
+        idx_dtype = get_index_dtype(maxval=self.shape[0] * self.shape[1])
```

Comment:
It looks like `get_index_dtype` was always casting to `np.int32` if the largest dimension in the array didn't overflow single-precision int storage for index generation. The new behavior encoded in this PR is that a set of indices over **all** the elements of an array (not just the rows or columns in isolation) has to fit in single precision int, otherwise `np.int64` will be used.

If I understand the linked issue discussion, the idea is to prevent a downstream sparse reshape operation from receiving an int32 dtype, which was working ok for the row / column indexing, but won't work for a reshape where the indices of all elements in the array may not fit in int32?

Reply:
`self.size` may be more concise (& clear?) than manually calculating the product here?

Reply:
> The new behavior encoded in this PR is that a set of indices over all the elements of an array (not just the rows or columns in isolation) has to fit in single precision int, otherwise `np.int64` will be used.

This is exactly correct.

> `self.size` may be more concise (& clear?) than manually calculating the product here?

You would think that `self.size` is the number of elements in the matrix, like `numpy.matrix`, but [it's not](https://github.com/scipy/scipy/blob/v1.1.0/scipy/sparse/base.py#L683-L684). It's the number of nonzero elements in the sparse array. This multiplication is the only way until we deprecate `size` and require people to use `nnz`, instead.